File: csharp/prism/Prism.cs
public static class Prism
{
    public readonly record struct LaserInfo(double X, double Y, double Angle);
    public readonly record struct PrismInfo(int Id, double X, double Y, double Angle);

    public static int[] FindSequence(LaserInfo laser, PrismInfo[] prisms)
    {
        var sequence = new List<int>();
        var currentX = laser.X;
        var currentY = laser.Y;
        var currentAngle = laser.Angle;

        for (int i = 0; i < 500; i++)
        {
            PrismInfo? nearest = null;
            var minDistance = double.MaxValue;

            // Calculate current direction vector once per hit
            var rad = currentAngle * (Math.PI / 180.0);
            var dxDir = Math.Cos(rad);
            var dyDir = Math.Sin(rad);

            foreach (var prism in prisms)
            {
                var dx = prism.X - currentX;
                var dy = prism.Y - currentY;
                var dist = Math.Sqrt(dx * dx + dy * dy);

                if (dist < 1e-7) continue;

                var pDx = dx / dist;
                var pDy = dy / dist;

                var dot = (pDx * dxDir) + (pDy * dyDir);

                if (dot > 0.999999) 
                {
                    if (dist < minDistance)
                    {
                        minDistance = dist;
                        nearest = prism;
                    }
                }
            }

            if (!nearest.HasValue) break;

            sequence.Add(nearest.Value.Id);
            currentX = nearest.Value.X;
            currentY = nearest.Value.Y;
            currentAngle += nearest.Value.Angle;
        }

        return [.. sequence];
    }
}

File: java/prism/src/main/java/Prism.java
import java.util.*;

public class Prism {

    public record LaserInfo(double x, double y, double angle) {
    }

    public record PrismInfo(int id, double x, double y, double angle) {
    }

    public static List<Integer> findSequence(LaserInfo laser, List<PrismInfo> prisms) {
        var sequence = new ArrayList<Integer>();
        var currentX = laser.x();
        var currentY = laser.y();
        var currentAngle = laser.angle();

        for (int i = 0; i < 500; i++) {
            PrismInfo nearest = null;
            var minDistance = Double.MAX_VALUE;

            var rad = currentAngle * (Math.PI / 180.0);
            var dxDir = Math.cos(rad);
            var dyDir = Math.sin(rad);

            for (var prism : prisms) {
                var dx = prism.x() - currentX;
                var dy = prism.y() - currentY;
                var dist = Math.sqrt(dx * dx + dy * dy);

                if (dist < 1e-7) continue;

                var pDx = dx / dist;
                var pDy = dy / dist;

                var dot = (pDx * dxDir) + (pDy * dyDir);

                if (dot > 0.999999 && dist < minDistance) {
                    minDistance = dist;
                    nearest = prism;
                }
            }

            if (nearest == null) break;

            sequence.add(nearest.id());
            currentX = nearest.x();
            currentY = nearest.y();
            currentAngle += nearest.angle();
        }

        return List.copyOf(sequence);
    }
}